### ISSUE TITLE
Proposal: add --hoistImports to compiler options

### ISSUE THREAD

**Issue by ericanderson** [Microsoft/TypeScript#22178]
In ES6, imports are hoisted and evaluated before code within the same file. [Webpack definitely already does this](https://github.com/webpack/webpack/issues/4120) and I think rollup might too (or certainly it would be legal for them to do so). 

Example:
```ts
import installPolyfills from "ie-polyfills";
installPolyfills();
import Foo from "other-code";
```

In ES6, your code will be evaluated as
```ts
import installPolyfills from "ie-polyfills";
import Foo from "other-code";
installPolyfills();
```

and cause an unexpected error (polyfills aren't available for `Foo`) since in an ES5 engine it still behaves as the example.

Unfortunately, this allows us to write code that doesn't behave the same depending on the output module type (if TS hoisted imports when `module: "es6"`) nor when running ES6 style code through other tools (like webpack or rollup).

This flag would instruct TS to hoist imports in generated code, regardless of the output module format. Thus one can write code that can be sure to behave the same in node 6, webpack, and engines that natively support es6 modules.

Thus, with `hostImports: true` and `module: "cjs"`, your output code would be:
```ts
var installPolyfills = require("ie-polyfills");
var Foo = require("other-code");
installPolyfills();
```

I think that ideally this flag would default to true when `module: "es6"` however that would likely be considered a breaking change. So perhaps that change (or always defaulting it to true, since TS looks like ES6) could happen in TS3.

Thanks
CC @mhegazy

**Comment by weswigham** [comment]
`module: "es6"` doesn't downlevel any imports - what starts as an `import` statement ends as an `import` statement. Do you mean `target`?

**Comment by ericanderson** [comment]
Thanks Wesley. Good catch. Updating

**Comment by weswigham** [comment]
It's also probably worth noting that is is the behavior for all module formats other than cjs.

**Comment by ericanderson** [comment]
Sorry, now that I am rereading what I wrote. When using webpack with `module: "es6"`, your imports are hoisted because it looks like es6 code.

**Comment by weswigham** [comment]
Right, but your point is that our behavior differs from spec when `module: "commonjs"` is set, yeah?

**Comment by ericanderson** [comment]
If TypeScript is a strict superset of es6, then the code generated when module: “cjs” is not to spec. 

Whether or not that’s what TypeScript is, I would like to ensure that my fellow engineers do not accidentally introduce bugs due to side effects when testing via node vs runtime environments.

**Comment by DanielRosenwasser** [comment]
Wait, is this #16166?

**Comment by ericanderson** [comment]
Yeah I guess it’s very similar. That’s more about having an error for out of order and this is an option to force hoisted imports even in cjs

**Comment by weswigham** [comment]
The resolution on the other issue was to commit to changing our cjs emit to hoist imports (not to issue some warning), which would is what we'd do here, too. As I said, it _is_ already how it behaves for every other target.
